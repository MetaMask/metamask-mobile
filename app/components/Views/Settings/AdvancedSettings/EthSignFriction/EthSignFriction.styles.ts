// Third party dependencies.
import { StyleSheet, TextStyle } from 'react-native';
import { typography } from '@metamask/design-tokens';
import { ThemeColors } from '@metamask/design-tokens/dist/js/themes/types';
/**
 * Style sheet function for EthSignFriction component.
 *
 * @returns StyleSheet object.
 */

export default (colors: ThemeColors) =>
  StyleSheet.create({
    frictionContainer: {
      alignItems: 'center',
      justifyContent: 'center',
      padding: 16,
      alignSelf: 'center',
    },
    heading: {
      ...(typography.sHeadingMD as TextStyle),
      color: colors.text.default,
    },
    explanationText: {
      ...(typography.sBodyMD as TextStyle),
      color: colors.text.default,
    },
    warningBox: {
      flexDirection: 'row',
      backgroundColor: colors.error.muted,
      borderLeftColor: colors.error.default,
      borderRadius: 4,
      borderLeftWidth: 4,
      marginTop: 24,
      paddingStart: 11,
      paddingEnd: 8,
      paddingVertical: 8,
    },
    warningText: {
      ...(typography.sBodyMD as TextStyle),
      color: colors.text.default,
      flex: 1,
      marginStart: 8,
    },
    important: {
      ...(typography.sBodyMDBold as TextStyle),
    },
    understandCheckboxView: {
      flexDirection: 'row',
      justifyContent: 'flex-start',
      margin: 16,
    },
    understandCheckboxLabel: {
      ...(typography.sBodyMD as TextStyle),
      color: colors.text.default,
      marginStart: 16,
    },
    iOnlySignInputView: {
      paddingVertical: 24,
    },
    iOnlySignInputLabel: {
      ...(typography.sBodyMDBold as TextStyle),
      color: colors.text.default,
      textAlign: 'left',
    },
    iOnlySignTextInput: {
      paddingVertical: 24,
      alignSelf: 'flex-start',
    },
    confirmTextError: {
      ...(typography.sBodySM as TextStyle),
      color: colors.error.default,
    },
    buttonsContainer: {
      flexDirection: 'row',
      justifyContent: 'space-evenly',
    },
    button: {
      flex: 1,
    },
    primaryButton: {
      marginStart: 16,
    },
  });
