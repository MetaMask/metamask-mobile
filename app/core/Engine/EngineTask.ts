import Realm from 'realm';

/**
 * Generic schema for persisting any controller state
 * Replaces FileSystemStorage key-value approach with Realm storage
 */
export class PersistedState extends Realm.Object<PersistedState> {
  key!: string;        // "persist:TransactionController", "persist:KeyringController", etc.
  data!: any;        // The actual controller state as JSON-serializable data
  updatedAt!: Date;    // When this controller state was last updated
  version!: number;    // Schema version for future migrations

  static schema: Realm.ObjectSchema = {
    name: 'PersistedState',
    primaryKey: 'key',
    properties: {
      key: 'string',
      data: 'mixed',
      updatedAt: 'date',
      version: { type: 'int', default: 1 },
    },
  };
}

// Keep the old Controller for backward compatibility during transition
export class Controller extends Realm.Object<Controller> {
  key!: string;
  value!: string;
  isActive!: boolean;

  static schema: Realm.ObjectSchema = {
    name: 'Controller',
    primaryKey: 'key',
    properties: {
      key: 'string',
      value: 'string',
      isActive: { type: 'bool', default: true },
    },
  };
}
