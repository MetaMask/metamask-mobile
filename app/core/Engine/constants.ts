/**
 * Messageable modules that are part of the Engine's context, but are not defined with state.
 * TODO: Replace with type guard once consistent inheritance for non-controllers is implemented. See: https://github.com/MetaMask/decisions/pull/41
 */
export const STATELESS_NON_CONTROLLER_NAMES = [
  'AssetsContractController',
  'NftDetectionController',
  'TokenDetectionController',
] as const;

export const BACKGROUND_STATE_CHANGE_EVENT_NAMES = [
  'AccountsController:stateChange',
  'AccountTrackerController:stateChange',
  'AddressBookController:stateChange',
  'ApprovalController:stateChange',
  'CurrencyRateController:stateChange',
  'GasFeeController:stateChange',
  'KeyringController:stateChange',
  'LoggingController:stateChange',
  'NetworkController:stateChange',
  'NftController:stateChange',
  'PermissionController:stateChange',
  'PhishingController:stateChange',
  'PPOMController:stateChange',
  'PreferencesController:stateChange',
  'RemoteFeatureFlagController:stateChange',
  'SelectedNetworkController:stateChange',
  'SignatureController:stateChange',
  'SmartTransactionsController:stateChange',
  'SwapsController:stateChange',
  'TokenBalancesController:stateChange',
  'TokenListController:stateChange',
  'TokenRatesController:stateChange',
  'TokensController:stateChange',
  'TransactionController:stateChange',
  ///: BEGIN:ONLY_INCLUDE_IF(preinstalled-snaps,external-snaps)
  'SnapController:stateChange',
  'SnapsRegistry:stateChange',
  'SubjectMetadataController:stateChange',
  'AuthenticationController:stateChange',
  'UserStorageController:stateChange',
  'NotificationServicesController:stateChange',
  'NotificationServicesPushController:stateChange',
  ///: END:ONLY_INCLUDE_IF
] as const;
