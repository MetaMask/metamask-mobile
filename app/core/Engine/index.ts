export { default } from './Engine';
export * from './types';
