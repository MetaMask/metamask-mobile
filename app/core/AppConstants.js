import Device from '../util/device';
import { MAINNET, RINKEBY } from '../constants/network';

const DEVELOPMENT = 'development';

export default {
  IS_DEV: process.env?.NODE_ENV === DEVELOPMENT,
  DEFAULT_LOCK_TIMEOUT: 30000,
  DEFAULT_SEARCH_ENGINE: 'DuckDuckGo',
  TX_CHECK_MAX_FREQUENCY: 5000,
  TX_CHECK_NORMAL_FREQUENCY: 10000,
  TX_CHECK_BACKGROUND_FREQUENCY: 30000,
  IPFS_OVERRIDE_PARAM: 'mm_override',
  IPFS_DEFAULT_GATEWAY_URL: 'https://cloudflare-ipfs.com/ipfs/',
  IPNS_DEFAULT_GATEWAY_URL: 'https://cloudflare-ipfs.com/ipns/',
  SWARM_DEFAULT_GATEWAY_URL: 'https://swarm-gateways.net/bzz:/',
  supportedTLDs: ['eth', 'xyz', 'test'],
  MAX_PUSH_NOTIFICATION_PROMPT_TIMES: 2,
  CONNEXT: {
    HUB_EXCHANGE_CEILING_TOKEN: 69,
    MIN_DEPOSIT_ETH: 0.03,
    MAX_DEPOSIT_TOKEN: 30,
    BLOCKED_DEPOSIT_DURATION_MINUTES: 5,
    CONTRACTS: {
      4: '0x0Fa90eC3AC3245112c6955d8F9DD74Ec9D599996',
      1: '0xdfa6edAe2EC0cF1d4A60542422724A48195A5071',
    },
    SUPPORTED_NETWORKS: [MAINNET, RINKEBY],
  },
  MM_UNIVERSAL_LINK_HOST: 'metamask.app.link',
  MM_DEEP_ITMS_APP_LINK: 'https://metamask.app.link/skAH3BaF99',
  SAI_ADDRESS: '0x89d24A6b4CcB1B6fAA2625fE562bDD9a23260359',
  HOMEPAGE_URL: process.env.MM_HOMEPAGE || 'https://home.metamask.io/',
  SHORT_HOMEPAGE_URL: 'MetaMask.io',
  ZERO_ADDRESS: '0x0000000000000000000000000000000000000000',
  USER_AGENT: Device.isAndroid()
    ? 'Mozilla/5.0 (Linux; Android 10; Android SDK built for x86 Build/OSM1.180201.023) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/77.0.3865.92 Mobile Safari/537.36'
    : 'Mozilla/5.0 (iPhone; CPU iPhone OS 13_1 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) CriOS/76.0.3809.123 Mobile/15E148 Safari/605.1',
  NOTIFICATION_NAMES: {
    accountsChanged: 'metamask_accountsChanged',
    unlockStateChanged: 'metamask_unlockStateChanged',
    chainChanged: 'metamask_chainChanged',
  },
  FIAT_ORDERS: {
    POLLING_FREQUENCY: 10000,
  },
  DEEPLINKS: {
    ORIGIN_DEEPLINK: 'deeplink',
    ORIGIN_QR_CODE: 'qr-code',
  },
  WALLET_CONNECT: {
    //One day in hours
    SESSION_LIFETIME: 24,
    LIMIT_SESSIONS: 20,
  },
  SWAPS: {
    ACTIVE: true,
    ONLY_MAINNET: true,
    CLIENT_ID: 'mobile',
    LIVENESS_POLLING_FREQUENCY: 5 * 60 * 1000,
    POLL_COUNT_LIMIT: 3,
    DEFAULT_SLIPPAGE: 2,
    CACHE_AGGREGATOR_METADATA_THRESHOLD: 5 * 60 * 1000,
    CACHE_TOKENS_THRESHOLD: 5 * 60 * 1000,
    CACHE_TOP_ASSETS_THRESHOLD: 5 * 60 * 1000,
  },
  MAX_SAFE_CHAIN_ID: 4503599627370476,
  URLS: {
    TERMS_AND_CONDITIONS: 'https://consensys.net/terms-of-use/',
    PRIVACY_POLICY: 'https://consensys.net/privacy-policy/',
    DATA_RETENTION_UPDATE:
      'https://consensys.net/blog/news/consensys-data-retention-update/',
    CONNECTIVITY_ISSUES:
      'https://metamask.zendesk.com/hc/en-us/articles/360059386712',
    NFT: 'https://metamask.zendesk.com/hc/en-us/articles/360058238591-NFT-tokens-in-MetaMask-wallet',
    SECURITY:
      'https://metamask.zendesk.com/hc/en-us/articles/360015489591-Basic-Safety-and-Security-Tips-for-MetaMask',
    TOKEN_BALANCE:
      'https://metamask.zendesk.com/hc/en-us/articles/360028059272-What-to-do-when-your-balance-of-ETH-and-or-ERC20-tokens-is-incorrect-inaccurate',
    MM_FAUCET: 'https://faucet.metamask.io/',
    WHY_TRANSACTION_TAKE_TIME:
      'https://community.metamask.io/t/what-is-gas-why-do-transactions-take-so-long/3172',
  },
  ERRORS: {
    INFURA_BLOCKED_MESSAGE:
      'EthQuery - RPC Error - This service is not available in your country',
  },
  GAS_OPTIONS: {
    LOW: 'low',
    MEDIUM: 'medium',
    HIGH: 'high',
    MARKET: 'market',
    AGGRESSIVE: 'aggressive',
  },
  GAS_TIMES: {
    UNKNOWN: 'unknown',
    MAYBE: 'maybe',
    LIKELY: 'likely',
    VERY_LIKELY: 'very_likely',
    AT_LEAST: 'at_least',
    LESS_THAN: 'less_than',
    RANGE: 'range',
  },
  REVIEW_PROMPT: {
    HIGH_GAS_FEES:
      'https://metamask.zendesk.com/hc/en-us/articles/360058751211-Why-my-gas-fees-are-so-high',
    MISSING_TOKENS:
      'https://metamask.zendesk.com/hc/en-us/articles/360015489031-How-to-add-unlisted-tokens-custom-tokens-in-MetaMask',
    SWAP_ISSUES:
      'https://metamask.zendesk.com/hc/en-us/articles/360060329612-Error-fetching-quote',
    SUPPORT: 'https://support.metamask.io',
  },
  BUNDLE_IDS: {
    IOS: 'io.metamask.MetaMask',
    ANDROID: 'io.metamask',
  },
  LEAST_SUPPORTED_ANDROID_API_LEVEL: 29,
  ADD_CUSTOM_NETWORK_POPULAR_TAB_ID: 'popular-tab',
  ADD_CUSTOM_NETWORK_CUSTOM_TAB_ID: 'custom-tab',
  REQUEST_SOURCES: {
    SDK_REMOTE_CONN: 'MetaMask-SDK-Remote-Conn',
    WC: 'WalletConnect',
    IN_APP_BROWSER: 'In-App-Browser',
  },
  MM_SDK: {
    SDK_CONNECTIONS: 'sdkConnections',
    SDK_APPROVEDHOSTS: 'sdkApprovedHosts',
  },
  CANCEL_RATE: 'Transactions (Cancel)',
  SPEED_UP_RATE: 'Transactions (Speed Up)',
  NETWORK_STATE_CHANGE_EVENT: 'NetworkController:stateChange',
  ETH_SIGN_ERROR: 'eth_sign requires 32 byte message hash',
};
