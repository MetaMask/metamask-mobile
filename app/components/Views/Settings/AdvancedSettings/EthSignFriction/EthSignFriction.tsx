// Third party dependencies.
import React, { useEffect, useMemo, useRef, useState } from 'react';
import { Platform, View } from 'react-native';
import { useNavigation } from '@react-navigation/native';

// External dependencies.
import SheetBottom, {
  SheetBottomRef,
} from '../../../../../component-library/components/Sheet/SheetBottom';
import { strings } from '../../../../../../locales/i18n';
import Text from '../../../../Base/Text';
import { useTheme } from '../../../../../util/theme';
import Icon, {
  IconColor,
  IconName,
  IconSize,
} from '../../../../../component-library/components/Icons/Icon';
import Engine from '../../../../../core/Engine';
import { tlc } from '../../../../../util/general';
import TextField from '../../../../../component-library/components/Form/TextField';
import Checkbox from '../../../../../component-library/components/Checkbox';
import Button, {
  ButtonSize,
  ButtonVariants,
  ButtonWidthTypes,
} from '../../../../../component-library/components/Buttons/Button';
import AppConstants from '../../../../../core/AppConstants';
import { trackEventV2 as trackEvent } from '../../../../../util/analyticsV2';
import { MetaMetricsEvents } from '../../../../../core/Analytics';

// Internal dependencies
import createStyles from './EthSignFriction.styles';
import generateTestId from '../../../../../../wdio/utils/generateTestId';
import {
  TOGGLE_ETH_SIGN_CONTINUE_BUTTON,
  TOGGLE_ETH_SIGN_MODAL,
  TOGGLE_ETH_SIGN_UNDERSTAND_CHECKBOX,
  TOGGLE_ETH_SIGN_UNDERSTAND_INPUT,
} from '../../../../../../wdio/screen-objects/testIDs/Components/ToggleEthSignModal.testIds';

/**
 * EthSignFriction Component.
 *
 * This component is used to show the friction bottomSheet for enabling eth_sign advanced setting.
 * The friction is divided into two steps:
 * - The first step is to show the user the risk of enabling eth_sign and requires a checkbox to continue.
 * - The second step is to ask the user to type a specific text to confirm that they understand the risk and allow them to enable eth_sign.
 */
const EthSignFriction = () => {
  const { colors, themeAppearance } = useTheme();
  const styles = createStyles(colors);
  const sheetRef = useRef<SheetBottomRef>(null);
  const [understandCheckbox, setUnderstandCheckbox] = useState(false);
  const [firstFrictionPassed, setFirstFrictionPassed] = useState(false);
  const [approveText, setApproveText] = useState<string>('');
  const navigation = useNavigation();

  // Track the view of the friction steps.
  useEffect(() => {
    if (!firstFrictionPassed) {
      trackEvent(
        MetaMetricsEvents.SETTINGS_ADVANCED_ETH_SIGN_FRICTION_FIRST_STEP_VIEWED,
        {},
      );
    } else {
      trackEvent(
        MetaMetricsEvents.SETTINGS_ADVANCED_ETH_SIGN_FRICTION_SECOND_STEP_VIEWED,
        {},
      );
    }
  }, [firstFrictionPassed]);

  // friction element status checks.
  const isApproveTextMatched = (text: string) =>
    tlc(text) === tlc(strings('app_settings.toggleEthSignModalFormValidation'));
  const isFirstStepReadyToContinue = useMemo(
    () => understandCheckbox && !firstFrictionPassed,
    [understandCheckbox, firstFrictionPassed],
  );
  const isSecondStepReadyToEnable = useMemo(
    () => firstFrictionPassed && isApproveTextMatched(approveText),
    [firstFrictionPassed, approveText],
  );
  const isPrimaryButtonDisabled = useMemo(
    () => !(isFirstStepReadyToContinue || isSecondStepReadyToEnable),
    [isFirstStepReadyToContinue, isSecondStepReadyToEnable],
  );

  // invert the current checkbox value: enabled -> disabled, disabled -> enabled.
  const toggleUnderstandCheckbox = () =>
    setUnderstandCheckbox(!understandCheckbox);

  // show a webview with the support page about eth_sign.
  const onLearnMorePress = () => {
    navigation.navigate('Webview', {
      screen: 'SimpleWebview',
      params: {
        url: AppConstants.URLS.WHAT_IS_ETH_SIGN_AND_WHY_IS_IT_A_RISK,
        title: strings('app_settings.enable_eth_sign'),
      },
    });
  };

  const onCancelPress = () => {
    sheetRef.current?.hide();
  };

  const onPrimaryPress = () => {
    if (isFirstStepReadyToContinue) {
      // when the user presses the primary "continue" button for the first step,
      // we mark first step as passed, and it shows the second friction step.
      setFirstFrictionPassed(true);
    } else if (isSecondStepReadyToEnable) {
      // when the user presses the primary "enable" button for the second step,
      // we enable the eth_sign advanced setting and hide the bottomSheet.
      const { PreferencesController } = Engine.context;
      PreferencesController.setDisabledRpcMethodPreference('eth_sign', true);
      trackEvent(MetaMetricsEvents.SETTINGS_ADVANCED_ETH_SIGN_ENABLED, {});
      sheetRef.current?.hide();
    }
  };

  return (
    <SheetBottom ref={sheetRef}>
      <View
        style={styles.frictionContainer}
        {...generateTestId(Platform, TOGGLE_ETH_SIGN_MODAL)}
      >
        {/*Common explanation content for both steps*/}
        <Icon
          name={IconName.Danger}
          color={IconColor.Error}
          size={IconSize.Xl}
        />
        <Text style={styles.heading}>
          {strings('app_settings.toggleEthSignModalTitle')}
        </Text>
        <Text style={styles.explanationText}>
          <Text>{strings('app_settings.toggleEthSignModalDescription')} </Text>
          <Text
            primary
            noMargin
            link
            accessibilityRole={'link'}
            accessible
            onPress={onLearnMorePress}
          >
            {strings('app_settings.toggleEthSignModalLearnMore')}
          </Text>
        </Text>
        <View style={styles.warningBox}>
          <Icon
            name={IconName.Danger}
            color={IconColor.Error}
            size={IconSize.Lg}
          />
          <Text style={styles.warningText}>
            {strings('app_settings.toggleEthSignModalBannerText')}
            <Text style={styles.important}>
              {strings('app_settings.toggleEthSignModalBannerBoldText')}
            </Text>
          </Text>
        </View>
        {!firstFrictionPassed ? (
          // First step checkbox content
          <View
            style={styles.understandCheckboxView}
            {...generateTestId(Platform, TOGGLE_ETH_SIGN_UNDERSTAND_CHECKBOX)}
          >
            <Checkbox
              isChecked={understandCheckbox}
              accessibilityRole={'checkbox'}
              accessible
              onPress={toggleUnderstandCheckbox}
            />
            <Text
              style={styles.understandCheckboxLabel}
              accessibilityRole={'button'}
              accessible
              onPress={toggleUnderstandCheckbox}
            >
              {strings('app_settings.toggleEthSignModalCheckBox')}
            </Text>
          </View>
        ) : (
          // Second step textfield content
          <View style={styles.iOnlySignInputView}>
            <Text style={styles.iOnlySignInputLabel}>
              {strings('app_settings.toggleEthSignModalFormLabel')}
            </Text>
            <TextField
              style={styles.iOnlySignTextInput}
              contextMenuHidden
              autoCompleteType={'off'}
              autoCorrect={false}
              returnKeyType={'done'}
              autoCapitalize="none"
              keyboardAppearance={themeAppearance}
              isError={approveText.length > 0 && isPrimaryButtonDisabled}
              accessibilityHint={strings(
                'app_settings.toggleEthSignModalFormLabel',
              )}
              onEndEditing={(e) => setApproveText(e.nativeEvent.text)}
              onFocus={() => setApproveText('')}
              {...generateTestId(Platform, TOGGLE_ETH_SIGN_UNDERSTAND_INPUT)}
            />
            {approveText.length > 0 && isPrimaryButtonDisabled && (
              <Text style={styles.confirmTextError}>
                {strings('app_settings.toggleEthSignModalFormError')}
              </Text>
            )}
          </View>
        )}
        {/*Common action buttons for both steps*/}
        <View style={styles.buttonsContainer}>
          <Button
            variant={ButtonVariants.Secondary}
            width={ButtonWidthTypes.Full}
            size={ButtonSize.Lg}
            style={styles.button}
            accessibilityRole={'button'}
            accessible
            label={strings('navigation.cancel')}
            onPress={onCancelPress}
          />
          <Button
            variant={ButtonVariants.Primary}
            width={ButtonWidthTypes.Full}
            size={ButtonSize.Lg}
            style={{ ...styles.button, ...styles.primaryButton }}
            isDanger={firstFrictionPassed}
            isDisabled={isPrimaryButtonDisabled}
            accessibilityState={{ disabled: isPrimaryButtonDisabled }}
            accessible
            accessibilityRole={'button'}
            label={strings(
              firstFrictionPassed
                ? 'app_settings.toggleEthSignEnableButton'
                : 'app_settings.toggleEthSignContinueButton',
            )}
            onPress={onPrimaryPress}
            {...generateTestId(Platform, TOGGLE_ETH_SIGN_CONTINUE_BUTTON)}
          />
        </View>
      </View>
    </SheetBottom>
  );
};

export default EthSignFriction;
