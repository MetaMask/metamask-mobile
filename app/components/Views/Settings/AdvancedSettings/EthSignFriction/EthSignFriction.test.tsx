import { shallow } from 'enzyme';
import React from 'react';
import EthSignFriction from './EthSignFriction';
import Engine from '../../../../../core/Engine';
import renderWithProvider from '../../../../../util/test/renderWithProvider';
import { fireEvent } from '@testing-library/react-native';
import { strings } from '../../../../../../locales/i18n';
import AppConstants from '../../../../../core/AppConstants';

const mockEngine = Engine;

jest.mock('../../../../../core/Engine', () => ({
  init: () => mockEngine.init({}),
}));

jest.mock('react-native-safe-area-context', () => {
  const inset = { top: 0, right: 0, bottom: 0, left: 0 };
  const frame = { width: 0, height: 0, x: 0, y: 0 };
  return {
    SafeAreaProvider: jest.fn().mockImplementation(({ children }) => children),
    SafeAreaConsumer: jest
      .fn()
      .mockImplementation(({ children }) => children(inset)),
    useSafeAreaInsets: jest.fn().mockImplementation(() => inset),
    useSafeAreaFrame: jest.fn().mockImplementation(() => frame),
  };
});

const mockNavigate = jest.fn();
const mockGoBack = jest.fn();
jest.mock('@react-navigation/native', () => {
  const actualNav = jest.requireActual('@react-navigation/native');
  return {
    ...actualNav,
    useNavigation: () => ({
      navigate: mockNavigate,
      goBack: mockGoBack,
    }),
  };
});

describe('Eth_sign friction bottom sheet', () => {
  it('should render initial friction step correctly', () => {
    const wrapper = shallow(<EthSignFriction />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should have checkbox and textfield unlocking steps', async () => {
    const {
      queryAllByText,
      queryAllByRole,
      getByRole,
      getByAccessibilityHint,
    } = renderWithProvider(<EthSignFriction />, {
      state: {},
    });

    // test if cancel button triggers navigation goBack
    const cancelButton = getByRole('button', {
      name: strings('navigation.cancel'),
    });
    fireEvent.press(cancelButton);
    expect(mockGoBack).toHaveBeenCalled();

    // test if readMore link triggers navigation to webview
    const learMoreLink = getByRole('link', {
      name: strings('app_settings.toggleEthSignModalLearnMore'),
    });
    fireEvent.press(learMoreLink);
    expect(mockNavigate).toHaveBeenCalledWith('Webview', {
      screen: 'SimpleWebview',
      params: {
        url: AppConstants.URLS.WHAT_IS_ETH_SIGN_AND_WHY_IS_IT_A_RISK,
        title: strings('app_settings.enable_eth_sign'),
      },
    });

    // expect the checkbox with associated label and continue button to be present
    expect(
      queryAllByText(strings('app_settings.toggleEthSignModalCheckBox')).length,
    ).toBe(1);

    const checkbox = getByRole('checkbox');

    const continueButton = getByRole('button', {
      name: strings('app_settings.toggleEthSignContinueButton'),
    });

    // Initially disabled continue button
    expect(continueButton.props.disabled).toBeTruthy();

    // Check the checkbox to enable continue button
    fireEvent.press(checkbox);
    expect(continueButton.props.disabled).toBeFalsy();

    // press continue button to move to step 2
    fireEvent.press(continueButton);

    // expect the continue button to be disabled again and text changed to 'enable'
    // as the bottom sheet changed to the second step
    const enableButton = getByRole('button', {
      name: strings('app_settings.toggleEthSignEnableButton'),
    });
    expect(enableButton.props.disabled).toBeTruthy();

    // Check the checkbox and continue button are gone
    expect(queryAllByRole('checkbox').length).toBe(0);
    expect(
      queryAllByRole('button', {
        name: strings('app_settings.toggleEthSignContinueButton'),
      }).length,
    ).toBe(0);

    // expect the textfield to be present
    const textField = getByAccessibilityHint(
      strings('app_settings.toggleEthSignModalFormLabel'),
    );

    // type the wrong text
    fireEvent(textField, 'onEndEditing', {
      nativeEvent: { text: 'not the right text' },
    });
    expect(enableButton.props.disabled).toBeTruthy();

    // type the write text
    fireEvent(textField, 'onEndEditing', {
      nativeEvent: { text: 'i only sign what i understand' },
    });
    expect(enableButton.props.disabled).toBeFalsy();
  });
});
