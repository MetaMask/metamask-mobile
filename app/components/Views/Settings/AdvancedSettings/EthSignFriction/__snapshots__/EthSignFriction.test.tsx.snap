// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Eth_sign friction bottom sheet should render initial friction step correctly 1`] = `
<ForwardRef>
  <View
    style={
      Object {
        "alignItems": "center",
        "alignSelf": "center",
        "justifyContent": "center",
        "padding": 16,
      }
    }
    testID="toggle-eth-sign-modal"
  >
    <Icon
      color="Error"
      name="Danger"
      size="32"
    />
    <Text
      style={
        Object {
          "color": "#24272A",
          "fontFamily": "Euclid Circular B",
          "fontSize": 18,
          "fontWeight": "700",
          "letterSpacing": 0,
          "lineHeight": 24,
        }
      }
    >
      Use at your own risk
    </Text>
    <Text
      style={
        Object {
          "color": "#24272A",
          "fontFamily": "Euclid Circular B",
          "fontSize": 14,
          "fontWeight": "400",
          "letterSpacing": 0,
          "lineHeight": 22,
        }
      }
    >
      <Text>
        Allowing eth_sign requests can make you vulnerable to phishing attacks. Always review the URL and be careful when signing messages that contain code.
         
      </Text>
      <Text
        accessibilityRole="link"
        accessible={true}
        link={true}
        noMargin={true}
        onPress={[Function]}
        primary={true}
      >
        Learn more.
      </Text>
    </Text>
    <View
      style={
        Object {
          "backgroundColor": "#D7384719",
          "borderLeftColor": "#D73847",
          "borderLeftWidth": 4,
          "borderRadius": 4,
          "flexDirection": "row",
          "marginTop": 24,
          "paddingEnd": 8,
          "paddingStart": 11,
          "paddingVertical": 8,
        }
      }
    >
      <Icon
        color="Error"
        name="Danger"
        size="24"
      />
      <Text
        style={
          Object {
            "color": "#24272A",
            "flex": 1,
            "fontFamily": "Euclid Circular B",
            "fontSize": 14,
            "fontWeight": "400",
            "letterSpacing": 0,
            "lineHeight": 22,
            "marginStart": 8,
          }
        }
      >
        If you've been asked to turn this setting on,
        <Text
          style={
            Object {
              "fontFamily": "Euclid Circular B",
              "fontSize": 14,
              "fontWeight": "700",
              "letterSpacing": 0,
              "lineHeight": 22,
            }
          }
        >
           you might be getting scammed.
        </Text>
      </Text>
    </View>
    <View
      style={
        Object {
          "flexDirection": "row",
          "justifyContent": "flex-start",
          "margin": 16,
        }
      }
      testID="toggle-eth-sign-understand-checkbox"
    >
      <Checkbox
        accessibilityRole="checkbox"
        accessible={true}
        isChecked={false}
        onPress={[Function]}
      />
      <Text
        accessibilityRole="button"
        accessible={true}
        onPress={[Function]}
        style={
          Object {
            "color": "#24272A",
            "fontFamily": "Euclid Circular B",
            "fontSize": 14,
            "fontWeight": "400",
            "letterSpacing": 0,
            "lineHeight": 22,
            "marginStart": 16,
          }
        }
      >
        I understand that I can lose all of my funds and NFTs if I enable eth_sign requests.
      </Text>
    </View>
    <View
      style={
        Object {
          "flexDirection": "row",
          "justifyContent": "space-evenly",
        }
      }
    >
      <Button
        accessibilityRole="button"
        accessible={true}
        label="Cancel"
        onPress={[Function]}
        size="48"
        style={
          Object {
            "flex": 1,
          }
        }
        variant="Secondary"
        width="full"
      />
      <Button
        accessibilityRole="button"
        accessibilityState={
          Object {
            "disabled": true,
          }
        }
        accessible={true}
        isDanger={false}
        isDisabled={true}
        label="Continue"
        onPress={[Function]}
        size="48"
        style={
          Object {
            "flex": 1,
            "marginStart": 16,
          }
        }
        testID="toggle-eth-sign-continue-button"
        variant="Primary"
        width="full"
      />
    </View>
  </View>
</ForwardRef>
`;
